src/audio/midi.rs

```diff
@@ -0,0 +1,146 @@
+    channel: Channel,
+    control_function: ControlFunction,
+) -> Option<ControlValue> {
+    messages.iter().rev().find_map(|message| match message {
+        MidiMessage::ControlChange(ch, function, value)
+            if ch == &channel && function == &control_function =>
+        {
+            Some(*value)
+        }
+
+        _ => None,
+    })
+}
+
+/// Maps a control value (0-127) into a range.
+/// eg. `control_value_in_range(50, 100, 64) == 75`
+pub fn control_value_in_range<T: Num + From<u8> + Copy>(min: T, max: T, value: ControlValue) -> T {
+    let value: T = from_control_value(value);
+    value * (max - min) / from_control_value(ControlValue::MAX) + min
+}
+
+fn midi_input() -> Result<MidiInput> {
+    Ok(MidiInput::new("Input")?)
+}
+
+fn port(name: &str) -> Result<MidiInputPort> {
+    let names = port_names()?;
+    let midi_input = midi_input()?;
+
+    midi_input
+        .ports()
+        .into_iter()
+        .find(|port| midi_input.port_name(&port) == Ok(name.into()))
+        .ok_or_else(|| {
+            anyhow!(
+                "Could not find a MIDI port with name '{}'. Available ports are:\n{}",
+                name,
+                names.join("\n")
+            )
+        })
+}
+
+fn handle_messages(port: MidiInputPort, sender: Sender<MidiMessage<'static>>) {
+    thread::spawn(move || {
+        // _connection needs to be a named parameter, because it needs to be kept alive until the end of the scope
+        let _connection = midi_input()
+            .unwrap()
+            .connect(
+                &port,
+                "midir-read-input",
+                move |_, message, _| {
+                    if let Some(message) =
+                        MidiMessage::try_from(message).unwrap().drop_unowned_sysex()
+                    {
+                        sender.send(message).unwrap();
+                    }
+                },
+                (),
+            )
+            .unwrap();
+
+        // keep this thread alive forever
+        thread::sleep(Duration::from_micros(u64::MAX));
+    });
+}
+
+fn from_control_value<T: From<u8>>(value: ControlValue) -> T {
+    let byte: u8 = value.into();
+    byte.into()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::convert::TryInto;
+
+    #[test]
+    fn test_from_control_value() {
+        assert_eq!(from_control_value::<u32>(0_u8.try_into().unwrap()), 0);
+        assert_eq!(from_control_value::<u32>(1_u8.try_into().unwrap()), 1);
+        assert_eq!(from_control_value::<u32>(127_u8.try_into().unwrap()), 127);
+    }
+
+    #[test]
+    fn test_control_value_in_range() {
+        assert_eq!(
+            control_value_in_range(0_u32, 100_u32, 0_u8.try_into().unwrap()),
+            0
+        );
+        assert_eq!(
+            control_value_in_range(0_u32, 100_u32, 127_u8.try_into().unwrap()),
+            100
+        );
+        assert_eq!(
+            control_value_in_range(0_u32, 100_u32, 64_u8.try_into().unwrap()),
```

Comment:
The name "in range" made me think that it was doing a min/max bounds cutoff but in fact this appears to be an ["interpolation"](https://docs.swmansion.com/react-native-reanimated/docs/1.x.x/nodes/interpolate/)

Reply:
Ah, thanks. I knew there was a better name for this.